Better API docs

Fishbowler opened the pull request:
Around half of the day-to-day functions I use were missing from the API docs. I've had a first crack at backfilling.

paulmelnikow commented on API.md:
```diff
@@ -1,64 +1,272 @@
+
+These are the commands you'll need to get IcedFrisby talking HTTP.
+
+An optional parameters object is accepted in all of these methods, but aren't _useful_ for every verb. All object parameters are optional.
```
"An optional `params` object is accepted by each of these methods" would be a little bit clearer. The params are useful for every verb, so I'd delete the second half of the first sentence.

Fishbowler replied:
With the exception of the mock, they're only useful for POST/PUT/PATCH, but you're right - it's disingenuous. I'll fix it.

paulmelnikow commented on API.md:
```diff
@@ -1,64 +1,272 @@
+### post(uri, data, params)
+
+- Types: `uri`: `string`, `data`:`object` `params`: `object` (optional)
+- Default: `none`
```
I realize this is the old format. In the future it would probably be clearer to list these one at a time, and indicate the default for each, and whether they are required:

- `uri`: The URI string.
- `data`: The data to post. An object when `{ json: true }` or `{ form: true }`, or else a data string / buffer.
- `params`: An optional params object, as described above.

Fishbowler replied:
I didn't like this format either, but kept consistent for this piece. 
If there's no love for it, I might change it now. 

Another alternative:

| Parameter | Description | Required |
| ----------- | ------------- | ----------- |
| uri | The URI string | Yes |
| data | The data to post. An object when { json: true } or { form: true }, or else a data string / buffer. | Yes |
| params | An optional params object, as described above. | No |

It's clean and clear, but is it heavyweight?

paulmelnikow replied:
I'm fine with that and feel like it's an improvement. You could also add a type column.

Fishbowler replied:
Done. I decided against the Type column since most of them were strings. Then wrote String almost everywhere anyway. We can revisit this later if we feel it's clunky. Maybe solicit some external feedback somehow?

paulmelnikow commented on API.md:
````diff
@@ -1,64 +1,272 @@
+
+```javascript
+frisby.create('a test')
+    .post('http://example.com/login', {
````
It might be slightly clearer to provide a more REST-like example. How about making this post to `https://example.com/account`?

Fishbowler replied:
I have no strong feelings on this. Updated.

paulmelnikow commented on API.md:
```diff
@@ -1,64 +1,272 @@
+    .get('http://example.com')
+    .addHeaders({
+      accept: 'text/html',
+      referer: 'http://www.test.net'
```
You could mention in a comment that the header field names are case insensitive, according to the HTTP spec?

Fishbowler replied:
Done.

paulmelnikow commented on API.md:
```diff
@@ -1,64 +1,272 @@
+
+### removeHeader(header)
+
+Removes a given header from the outgoing request
```
It's not really clear here why this would be useful. Could you clarify that?

Fishbowler replied:
I wouldn't like to pre-suppose what others might want to do with this functionality. The example I gave is where Content-Type is set by us, but the test developer doesn't want to send it, perhaps to test that their API is resilient to such things. I'd prefer to keep it in the docs unless we remove the function or rename it to private.

paulmelnikow replied:
I think it would be helpful to write some of what you just said into the docs! The docs should definitely tell a developer "what this does," but it's also really helpful to tell a developer "what can I do with this" / "why I might want to use this."

Fishbowler replied:
Done!

paulmelnikow commented on API.md:
````diff
@@ -503,36 +752,119 @@ frisby.create('First test')
+    .toss()
+```
+
+This function can also be called with no parameter to return the current configured timeout (either by default, by global setup or having used this function with a parameter previously).
````
This is consistent with the current behavior, though it makes me wonder, do we need to do this? ^^

Fishbowler replied:
I found it odd. I can't see the use case from the test developer.

paulmelnikow replied:
Agreed. Maybe we can remove it before the 2.0.0 release.

paulmelnikow commented on API.md:
```diff
@@ -125,10 +337,11 @@ frisby.create('Ensure response returns one cookie called "auth"')
-* Types: `path`: `string`, `json`: `JSON`
-* Defaults: `none`
+- Types: `path`: `string`, `json`: `JSON`
+- Defaults: `none`
```
It's worth mentioning what `path` is, or else moving the "Using Paths" section before the first one of these.

Fishbowler replied:
Added an extra bit to each method pointing the user at the Using Paths section.

paulmelnikow commented on API.md:
```diff
@@ -503,36 +752,119 @@ frisby.create('First test')
+    .after(function(err, res, body, headers) {
+        frisby.create('Check Item Exists')
+            .get('http://example.com/users/list')
+            .waits(500) //Allow time for extra server-side processing of the user
```
This comment could be moved to the method comment:

> Sets a period of time in milliseconds to wait after `toss()` before the request is sent. This can allow time for server-side processing between chained requests.

Fishbowler replied:
🤦‍♂️  Yeah. Fixed.

paulmelnikow commented on API.md:
```diff
@@ -503,36 +752,119 @@ frisby.create('First test')
+
+### exceptionHandler(function)
+
+Sets a function to run if an error is raised. Can be used to output additional debug info not covered by the [inspectors](#inspectors), or perhaps to add validation to a non-deterministic result.
```
Just curious, do you use this anywhere? It's a behavior I'm not sure we need. What you're saying about additional debug info makes sense, though it seems like we could do that in `inspect`. You mention validation for a non-determinstic result, and I'm not sure I understand that. I'm kind of puzzled.

Fishbowler replied:
My only usage so far has been in IcedFrisby unit tests, where it's got lots of usage. That said, I can imagine a case where I want to assert that the response for "pending notifications", an endpoint that for "reasons" is immune to Arrange/Act/Assert, is either empty or matches a specific format. I could use exceptionHandler for this.

Not sure I entirely believe this use case fully myself (because if we wanted to support "either A or B" expects, we could do a much better job), but I definitely wanted to include something about this method as it's publicly available. An alternative could be renaming it to `_exceptionHandler`.

paulmelnikow replied:
Couldn't this use case be handled more cleanly using `.after()`?

> … I definitely wanted to include something about this method as it's publicly available. An alternative could be renaming it to `_exceptionHandler`.

I'm fine with including it for the moment, though like your idea of renaming it.

paulmelnikow commented on API.md:
````diff
@@ -566,6 +902,7 @@ frisby.create('Just a quick inspection of the JSON HTTP response')
 
 Console output:
+
 ```json
````
This won't be screaming so loud if you remove the `json` here.

Fishbowler commented:
I've added all the changes from the feedback, but I've also added a little more info as part of the final tables work. Can you read over that final bit of info @paulmelnikow, just for accuracy (esp L808 and L827)?

paulmelnikow commented on API.md:
```diff
@@ -503,36 +783,132 @@ frisby.create('First test')
+
+| Parameter | Description | Required |
+| --------- | ----------- | -------- |
+| ms | Integer. Sets the mocha timeout for this request. | Yes |
```
Ah, this is incorrect. The `timeout()` is the timeout per request, which is compounded into a longer mocha timeout based on `delay()` and `retries()`. When the total time is reached, the test fails.

Fishbowler replied:
Fixed.

paulmelnikow commented on API.md:
```diff
@@ -503,36 +783,132 @@ frisby.create('First test')
+
+This function can also be called with no parameter to return the current configured timeout (either by default, by global setup or having used this function with a parameter previously).
+
+This is different from [expectMaxResponseTime](#expectmaxresponsetimems) in that expectMaxResponseTime will await the request to complete then test the amount of time that the request took (after any preceeding expect functions). This will abort the request and so not run any developer-defined expect functions on the response.
```
> This will abort the request and so not run any developer-defined expect functions on the response.

This is not guaranteed. At least I think it's not guaranteed.

How about this?

> When a timeout occurs, the test will be aborted. The expectations and inspections may or may not run and will not be printed. If you want to run all the expectations and inspections even when the response is slow, use `expectMaxResponseTime` instead.

Fishbowler replied:
I think I got that nugget from the comments on the method, but your version is safer and snappier.

Note to self: Test this.

paulmelnikow commented on API.md:
````diff
@@ -503,36 +783,132 @@ frisby.create('First test')
+    .toss()
+```
+
+In this above example, the first request would wait 2000ms for a response (the default), then the first retry would begin 250ms later, allow another 2000ms for this to perhaps complete, then wait 500ms (2 x 250ms) before beginning the second retry.
````
> This makes two retries, for a total of three attempts, waiting 250 ms between them. For each one, the timeout resets to 5000 ms (the default).

Fishbowler replied:
Thanks! Fixed.

paulmelnikow commented on API.md:
```diff
@@ -1,64 +1,283 @@
+
+### toss()
+
+Used to complete the list of commands and issue the request.
```
> Complete the list of commands and register the test with Mocha.

Fishbowler replied:
Much better. I knew that wasn't quite accurate when I wrote it, but couldn't word it better.

paulmelnikow commented on API.md:
```diff
@@ -1,64 +1,283 @@
+| Parameter | Description | Required |
+| --------- | ----------- | -------- |
+| header | String giving the name of the header. Case insensitive. | Yes |
+| content | String giving the value of the given header. Case sensitive. | Yes |
```
Note that you can invoke this multiple times per header?

Fishbowler replied:
Done.

paulmelnikow commented on API.md:
```diff
@@ -224,14 +473,28 @@ frisby.create('Ensure response arrives within two seconds')
 
+### not()
+
+Negates an expect, inverting the logic to expect the opposite, e.g. JSON doesn't match.
```
What happens if there are more than one `expect`?

Fishbowler replied:
Ermm. Dunno...

Reading the code, it sets the mode, so would apply to all expects in this chain. I say all expects. Looking at usages, it's just pathMatch ones - `expectJSON`, `expectJSONTypes`, `expectContainsJSON` and `expectJSONLength`. Won't apply to the other inbuilt ones, or to plugins that don't have explicit support.

I've updated the docs to match, but we might want to revisit this and make it more sturdy at some point.

paulmelnikow replied:
Sounds great.

paulmelnikow commented on API.md:
```diff
@@ -503,36 +783,132 @@ frisby.create('First test')
+
+### waits(ms)
+
+Sets a period of time in milliseconds to wait after `toss()` before the request is sent. This can allow time for server-side processing between chained requests.
```
The wait is between the start of the test and when the request is sent. The call to `toss()` could have occurred earlier. The wait happens after the `before()` hooks.

Fishbowler replied:
Thanks! Fixed!

paulmelnikow commented:
I took another pass through and left you more comments. Great work on this!

Fishbowler commented:
Thanks Paul! I've addressed them all, I believe. Not sure on how to deal with the weak `not()`. Gut feel is to create a new issue.

paulmelnikow commented:
New issue sounds good to me.

Fishbowler commented:
Raised `not()` problem as #95

paulmelnikow commented:
Glad to have this in!

Fishbowler commented:
Hmmm. Seems I should have squashed & merged rather than merged to stay consisten, Has any other work gone on? Can I do a sneaky force push to fix this @paulmelnikow?

paulmelnikow commented:
Sure, I'm fine with that! I do think a straight-line history is easier to follow.

Fishbowler commented:
Nailed it,.
